File: pharmacy-search/backend/app/checkout.py
import time

from fastapi import APIRouter, Depends, HTTPException
from pydantic import BaseModel, Field

from app.auth import get_current_user
from app.cart import _get_cart
from app.db import get_connection, row_to_dict, write_transaction

router = APIRouter(prefix="/api/checkout", tags=["checkout"])

# An order holding a prescription-only medicine does not enter fulfilment on
# the customer's say-so. It parks here until a pharmacist reviews it, which is
# an order-level hold rather than a cart-level block: the shopper can still
# check out, they just cannot be shipped Rx items without review.
STATUS_AWAITING_PRESCRIPTION = "awaiting_prescription"

SHIPPING_METHODS = {
    "standard": {"label": "Standard (3-5 days)", "fee": 0},
    "express": {"label": "Express (1-2 days)", "fee": 30000},
}

# Only COD is real. The others are shown so the checkout flow looks and feels
# complete, but selecting one does not charge anyone or contact any payment
# processor - there is no VNPay/Momo/Stripe integration wired up. Building a
# card-entry form here would be actively wrong: real payment processing
# should go through one of those processors, not be simulated in-house.
PAYMENT_METHODS = {
    "cod": {"label": "Cash on delivery", "available": True},
    "vnpay": {"label": "VNPay", "available": False},
    "momo": {"label": "Momo", "available": False},
}


@router.get("/options")
def get_options():
    return {"shipping_methods": SHIPPING_METHODS, "payment_methods": PAYMENT_METHODS}


def _serialize_order(conn, order_row: dict) -> dict:
    # LEFT JOIN, not INNER: order_items must still render if the product was
    # later deactivated (soft-delete keeps the row) - image_url just comes
    # back None in that edge case instead of dropping the line item.
    items = conn.execute(
        """SELECT order_items.*, products.image_url FROM order_items
           LEFT JOIN products ON products.id = order_items.product_id
           WHERE order_items.order_id = ?""",
        (order_row["id"],),
    ).fetchall()
    return {
        "id": order_row["id"],
        "status": order_row["status"],
        "shipping_method": order_row["shipping_method"],
        "shipping_fee": order_row["shipping_fee"],
        "payment_method": order_row["payment_method"],
        "subtotal": order_row["subtotal"],
        "discount_code": order_row["discount_code"],
        "discount_amount": order_row["discount_amount"],
        "total": order_row["total"],
        "created_at": order_row["created_at"],
        "requires_prescription": bool(order_row["requires_prescription"]),
        "prescription_status": order_row["prescription_status"],
        "prescription_reference": order_row["prescription_reference"],
        "items": [
            {
                "product_id": i["product_id"],
                "webName": i["web_name"],
                "imageUrl": i["image_url"],
                "quantity": i["quantity"],
                "unit_price": i["unit_price"],
                "price_was_estimated": bool(i["price_was_estimated"]),
                "line_total": i["line_total"],
            }
            for i in items
        ],
    }


@router.get("/orders")
def list_orders(user: dict = Depends(get_current_user)):
    with get_connection() as conn:
        rows = conn.execute(
            "SELECT * FROM orders WHERE user_id = ? ORDER BY created_at DESC", (user["id"],)
        ).fetchall()
        return [_serialize_order(conn, row_to_dict(r)) for r in rows]


@router.get("/orders/{order_id}")
def get_order(order_id: int, user: dict = Depends(get_current_user)):
    with get_connection() as conn:
        row = conn.execute(
            "SELECT * FROM orders WHERE id = ? AND user_id = ?", (order_id, user["id"])
        ).fetchone()
        if not row:
            raise HTTPException(status_code=404, detail="Order not found")
        return _serialize_order(conn, row_to_dict(row))


class PlaceOrderRequest(BaseModel):
    address_id: int
    shipping_method: str
    payment_method: str
    # Identifier for the prescription the customer is presenting - today a
    # reference string (a clinic's document number, or an upload id once
    # object storage exists). Required whenever the cart holds an Rx item.
    # This is not the verification itself; the pharmacist review below is.
    prescription_reference: str | None = Field(default=None, max_length=200)


@router.post("/place-order", status_code=201)
def place_order(body: PlaceOrderRequest, user: dict = Depends(get_current_user)):
    if body.shipping_method not in SHIPPING_METHODS:
        raise HTTPException(status_code=400, detail="Unknown shipping method")
    if body.payment_method not in PAYMENT_METHODS:
        raise HTTPException(status_code=400, detail="Unknown payment method")

    # BEGIN IMMEDIATE, not the default deferred transaction: everything from
    # the stock read to the decrement has to be one serialized unit, or two
    # checkouts interleave and both sell the last box.
    with write_transaction() as conn:
        address = conn.execute(
            "SELECT id FROM addresses WHERE id = ? AND user_id = ?", (body.address_id, user["id"])
        ).fetchone()
        if not address:
            raise HTTPException(status_code=404, detail="Shipping address not found")

        cart = _get_cart(conn, user["id"])
        if not cart["items"]:
            raise HTTPException(status_code=400, detail="Your cart is empty")

        # Prescription-only medicines cannot be dispensed on a click. The cart
        # is allowed to contain them; the order is held for pharmacist review
        # before it can ship, and it cannot be created at all without a
        # prescription reference to review.
        requires_prescription = bool(cart["requires_prescription"])
        prescription_reference = (body.prescription_reference or "").strip()
        if requires_prescription and not prescription_reference:
            raise HTTPException(
                status_code=400,
                detail={
                    "message": "A prescription is required for these items before this order can be placed.",
                    "prescription_items": cart["prescription_items"],
                },
            )

        shipping_fee = SHIPPING_METHODS[body.shipping_method]["fee"]
        subtotal = cart["subtotal"]
        discount_amount = cart["discount_amount"]
        total = subtotal - discount_amount + shipping_fee
        if requires_prescription:
            status = STATUS_AWAITING_PRESCRIPTION
        else:
            status = "placed" if body.payment_method == "cod" else "pending_payment"
        now = time.time()

        cursor = conn.execute(
            """INSERT INTO orders
            (user_id, address_id, shipping_method, shipping_fee, payment_method,
             subtotal, discount_code, discount_amount, total, status, created_at,
             requires_prescription, prescription_status, prescription_reference)
            VALUES (?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?)""",
            (
                user["id"], body.address_id, body.shipping_method, shipping_fee, body.payment_method,
                subtotal, cart["discount_code"], discount_amount, total, status, now,
                int(requires_prescription),
                "pending_review" if requires_prescription else None,
                prescription_reference or None,
            ),
        )
        order_id = cursor.lastrowid

        for line in cart["items"]:
            # The decrement *is* the stock check. A read-then-write pair let
            # two orders both pass validation and both subtract, driving stock
            # negative; a conditional UPDATE that affects zero rows is how the
            # loser of that race finds out, and NULL stock (never set by the
            # crawl importer, but permitted by the schema) fails it too rather
            # than raising a TypeError halfway through the order.
            updated = conn.execute(
                "UPDATE products SET stock = stock - ? WHERE id = ? AND is_active = 1 "
                "AND stock IS NOT NULL AND stock >= ?",
                (line["quantity"], line["product_id"], line["quantity"]),
            )
            if updated.rowcount == 0:
                current = conn.execute(
                    "SELECT stock, is_active FROM products WHERE id = ?", (line["product_id"],)
                ).fetchone()
                if not current or not current["is_active"]:
                    raise HTTPException(status_code=409, detail=f"{line['webName']} is no longer available")
                available = current["stock"] or 0
                raise HTTPException(
                    status_code=409,
                    detail=f"Only {available} left of {line['webName']} - update your cart quantity",
                )
            conn.execute(
                """INSERT INTO order_items (order_id, product_id, web_name, quantity, unit_price, price_was_estimated, line_total)
                VALUES (?, ?, ?, ?, ?, ?, ?)""",
                (order_id, line["product_id"], line["webName"], line["quantity"], line["price"],
                 int(line["price_is_estimated"]), line["line_total"]),
            )
            conn.execute("DELETE FROM cart_items WHERE user_id = ? AND product_id = ?", (user["id"], line["product_id"]))

        conn.execute("DELETE FROM cart_discounts WHERE user_id = ?", (user["id"],))

        order_row = conn.execute("SELECT * FROM orders WHERE id = ?", (order_id,)).fetchone()
        return _serialize_order(conn, row_to_dict(order_row))

File: pharmacy-search/frontend/src/views/checkout.js
import { getCart, listAddresses, getCheckoutOptions, placeOrder, getToken } from '../api.js';
import { escapeHtml, formatVnd } from '../shared.js';
import { t } from '../i18n.js';
import { navigate } from '../router.js';
import { refreshNavBadges } from '../nav.js';

export async function render(root) {
  if (!getToken()) {
    root.innerHTML = `<div class="empty-state">${t('common.signInTo.checkout')}</div>`;
    return;
  }

  root.innerHTML = `<div class="loading">${t('common.loading')}</div>`;
  const [cart, addresses, options] = await Promise.all([getCart(), listAddresses(), getCheckoutOptions()]);

  if (!cart.items.length) {
    root.innerHTML = `<div class="empty-state">${t('checkout.emptyCart')}</div>`;
    return;
  }
  if (!addresses.length) {
    root.innerHTML = `<div class="empty-state">${t('checkout.needAddress')}</div>`;
    return;
  }

  const state = {
    addressId: addresses.find((a) => a.is_default)?.id ?? addresses[0].id,
    shippingMethod: Object.keys(options.shipping_methods)[0],
    paymentMethod: 'cod',
    prescriptionReference: '',
  };

  // Prescription-only items do not block the cart, but they do hold the
  // order: it is created in 'awaiting_prescription' and a pharmacist has to
  // review it before anything ships.
  const rxSection = cart.requires_prescription
    ? `
        <section class="checkout-section checkout-rx">
          <h2>${t('checkout.rxTitle')}</h2>
          <p class="hint-text">${t('checkout.rxNote', { items: cart.prescription_items.map(escapeHtml).join(', ') })}</p>
          <label>${t('checkout.rxReference')}
            <input type="text" id="rx-reference" maxlength="200" placeholder="${t('checkout.rxPlaceholder')}" required />
          </label>
        </section>`
    : '';

  root.innerHTML = `
    <h1 class="page-title">${t('checkout.title')}</h1>
    <div class="checkout-layout">
      <div class="checkout-form">
        <section class="checkout-section">
          <h2>${t('checkout.address')}</h2>
          <div class="radio-list">
            ${addresses
              .map(
                (a) => `
              <label class="radio-option">
                <input type="radio" name="address" value="${a.id}" ${a.id === state.addressId ? 'checked' : ''} />
                <span><strong>${escapeHtml(a.label)}</strong> — ${escapeHtml(a.recipient_name)}, ${escapeHtml(a.phone)}<br/>
                <span class="hint-text">${escapeHtml(a.line1)}, ${escapeHtml(a.city)}</span></span>
              </label>`
              )
              .join('')}
          </div>
        </section>

        <section class="checkout-section">
          <h2>${t('checkout.shipping')}</h2>
          <div class="radio-list">
            ${Object.entries(options.shipping_methods)
              .map(
                ([key, m]) => `
              <label class="radio-option">
                <input type="radio" name="shipping" value="${key}" ${key === state.shippingMethod ? 'checked' : ''} />
                <span>${escapeHtml(m.label)} — ${m.fee > 0 ? formatVnd(m.fee) : t('common.free')}</span>
              </label>`
              )
              .join('')}
          </div>
        </section>

        ${rxSection}

        <section class="checkout-section">
          <h2>${t('checkout.payment')}</h2>
          <div class="radio-list">
            ${Object.entries(options.payment_methods)
              .map(
                ([key, m]) => `
              <label class="radio-option">
                <input type="radio" name="payment" value="${key}" ${key === state.paymentMethod ? 'checked' : ''} />
                <span>${escapeHtml(m.label)} ${!m.available ? `<span class="estimate-tag">${t('checkout.notConnected')}</span>` : ''}</span>
              </label>`
              )
              .join('')}
          </div>
          <p class="hint-text">${t('checkout.paymentNote')}</p>
        </section>
      </div>

      <aside class="cart-summary">
        <h2>${t('cart.summary')}</h2>
        <div id="checkout-lines"></div>
        <div class="summary-row"><span>${t('cart.subtotal')}</span><span>${formatVnd(cart.subtotal) || formatVnd(0)}</span></div>
        ${cart.discount_code ? `<div class="summary-row summary-discount"><span>${t('cart.discount', { code: escapeHtml(cart.discount_code) })}</span><span>−${formatVnd(cart.discount_amount)}</span></div>` : ''}
        <div class="summary-row" id="shipping-fee-row"></div>
        <div class="summary-row summary-total" id="grand-total-row"></div>
        <div id="place-order-error" class="form-error" role="alert"></div>
        <button type="button" id="place-order-btn">${cart.requires_prescription ? t('checkout.placeOrderForReview') : t('checkout.placeOrder')}</button>
      </aside>
    </div>
  `;

  root.querySelector('#checkout-lines').innerHTML = cart.items
    .map((item) => `<div class="summary-row"><span>${escapeHtml(item.webName)} × ${item.quantity}</span><span>${formatVnd(item.line_total)}</span></div>`)
    .join('');

  function updateTotals() {
    const fee = options.shipping_methods[state.shippingMethod].fee;
    root.querySelector('#shipping-fee-row').innerHTML = `<span>${t('checkout.shippingFee')}</span><span>${fee > 0 ? formatVnd(fee) : t('common.free')}</span>`;
    const total = cart.subtotal - cart.discount_amount + fee;
    root.querySelector('#grand-total-row').innerHTML = `<span>${t('cart.total')}</span><span>${formatVnd(total)}</span>`;
  }
  updateTotals();

  root.querySelectorAll('input[name="address"]').forEach((r) => r.addEventListener('change', (e) => (state.addressId = parseInt(e.target.value, 10))));
  root.querySelectorAll('input[name="shipping"]').forEach((r) => r.addEventListener('change', (e) => { state.shippingMethod = e.target.value; updateTotals(); }));
  root.querySelectorAll('input[name="payment"]').forEach((r) => r.addEventListener('change', (e) => (state.paymentMethod = e.target.value)));
  root.querySelector('#rx-reference')?.addEventListener('input', (e) => (state.prescriptionReference = e.target.value.trim()));

  root.querySelector('#place-order-btn').addEventListener('click', async () => {
    const btn = root.querySelector('#place-order-btn');
    const errorEl = root.querySelector('#place-order-error');
    errorEl.textContent = '';
    btn.disabled = true;
    btn.textContent = t('checkout.placing');
    try {
      const order = await placeOrder({
        addressId: state.addressId,
        shippingMethod: state.shippingMethod,
        paymentMethod: state.paymentMethod,
        prescriptionReference: state.prescriptionReference,
      });
      refreshNavBadges();
      navigate(`#/order/${order.id}`);
    } catch (err) {
      errorEl.textContent = err.message;
      btn.disabled = false;
      btn.textContent = cart.requires_prescription ? t('checkout.placeOrderForReview') : t('checkout.placeOrder');
    }
  });
}
